src/fhirdefs/load.ts

```diff
@@ -1,29 +1,113 @@
-// TODO: Load from package files instead of these static folders.
 import { FHIRDefinitions } from './FHIRDefinitions';
-import fs from 'fs';
+import { PackageLoadError, DevPackageLoadError } from '../errors';
+import fs from 'fs-extra';
+import path from 'path';
+import os from 'os';
+import tar from 'tar';
+import rp from 'request-promise-native';
+import temp from 'temp';
 
-const _cache: Map<string, FHIRDefinitions> = new Map();
+/**
+ * Loads a dependency from user FHIR cache or from online
+ * @param {string} packageName - The name of the package to load
+ * @param {string} version - The version of the package to load
+ * @param {FHIRDefinitions} FHIRDefs - The FHIRDefinitions to load the dependencies into
+ * @param {string} cachePath - The path to load the package into
+ * @returns {Promise<FHIRDefinitions>} the loaded FHIRDefs
+ * @throws {PackageLoadError} when the desired package can't be loaded
+ */
+export async function loadDependency(
+  packageName: string,
+  version: string,
+  FHIRDefs: FHIRDefinitions,
+  cachePath: string = path.join(os.homedir(), '.fhir', 'packages')
+): Promise<FHIRDefinitions> {
+  const fullPackageName = `${packageName}#${version}`;
+  const loadPath = path.join(cachePath, fullPackageName, 'package');
+  let loadedPackage: string;
+  if (version != 'current') {
+    loadedPackage = loadFromPath(loadPath, fullPackageName, FHIRDefs);
+  }
+  if (!loadedPackage) {
+    let packageUrl: string;
+    if (version === 'dev') {
+      // Dev packages must be present in local FHIR cache
+      throw new DevPackageLoadError(fullPackageName);
+    } else if (version === 'current') {
+      // Current packages need to be loaded using build.fhir.org
+      const baseUrl = 'http://build.fhir.org/ig';
+      const res = await rp.get({ uri: `${baseUrl}/qas.json`, encoding: null });
+      const decoded = String.fromCharCode(...new Uint16Array(res));
+      const qaJSON = JSON.parse(decoded);
+      // Find matching packages and sort by date to get the most recent
+      const matchingPackages = qaJSON.filter((p: any) => p['package-id'] === packageName);
+      const newestPackage = matchingPackages.sort((p1: any, p2: any) => {
+        return Date.parse(p2['date']) - Date.parse(p1['date']);
+      })[0];
+      // Current packages are stored at build.fhir.org
+      packageUrl = `${baseUrl}/${newestPackage.repo}/package.tgz`;
+    } else {
+      // All non-current packages are stored at packages.fhir.org
+      packageUrl = `http://packages.fhir.org/${packageName}/${version}`;
+    }
+    // Create a temporary file and write the package to there
+    temp.track();
+    const tempFile = temp.openSync(); // TODO CHANGE THIS TO USE OTHER TEMP LIB
```

Comment:
Is this using the correct temp library now? If so, you can remove the TODO.

Reply:
Yes it's using `temp` not `tmp` now, I'll remove that.